sgo/types/lookup.go

```diff
@@ -294,6 +293,36 @@ func assertableTo(V *Interface, T Type) (method *Func, wrongType bool, mustOptio
 	return
 }
 
+// typeMustOptional reports whether a type T recursively needs to be wrapped in an optional.
+// If T is Array or Slice, the underlying type will be checked. If T is Optional, it will not
+// need to be wrapped in an optional unless the underlying type is a Map. In that case, further
+// checks are needed for the key and value type of the map.
+func typeMustOptional(T Type) bool {
+	t := T.Underlying()
+	switch t.(type) {
+	case *Pointer, *Map, *Signature, *Chan, *Interface:
+		return true
+	case *Slice, *Array:
```

Comment:
It's not the struct that must be wrapped, it's the _fields_ (that have an "optionable" type) which should be checked, just like you're checking the _elements_ of slices and arrays (and what's an array of length N if not syntactic sugar for a struct of N fields?). Examples of things that should be forbidden:

```go
foo.(struct{Bar *Bar)} // should be ?*Bar
foo.(struct{Err error, Bars []*Bar)} // should be ?error, []?*Bar
```

Plus, I just noticed that you're checking the underlying instead of the type itself. You shouldn't, because that will make you trespass named types (e. g. that `Foo` up there), and, as we talked about, this check must be made only for type-asserting to anonymous types (e. g. `foo.(struct{Bar *Bar})`), because SGo will guarantee that named types are constructed right.

Your example won't compile, because `Foo` isn't a pointer. If we remove the optional wrap:

```go
type Foo struct {
    Bar *Bar
}

// ...
myFoo := foo.(Foo)
myFoo.Bar.DoSomething() // This will crash
```

That will never crash, because if `foo` is a `Foo`, it must have been constructed as a `Foo{Bar: x}` somewhere, and SGo will enforce that `x` in that expression is not nil.